vhost-user-backend/src/handler.rs

```diff
@@ -712,6 +716,44 @@ where
         self.uffd = None;
         Ok(())
     }
+
+    // Sets logging (i.e., bitmap) shared memory space.
+    //
+    // During live migration, the front-end may need to track the modifications the back-end
+    // makes to the memory mapped regions. The front-end should mark the dirty pages in a log.
+    // Once it complies to this logging, it may declare the `VHOST_F_LOG_ALL` vhost feature.
+    //
+    // If the backend has the `VHOST_USER_PROTOCOL_F_LOG_SHMFD` protocol feature it may receive
+    // the `VHOST_USER_SET_LOG_BASE` message. The log memory file descriptor is provided in `file`,
+    // the size and offset of shared memory area are provided in the `VhostUserLog` message.
+    //
+    // See https://qemu-project.gitlab.io/qemu/interop/vhost-user.html#migration.
+    fn set_log_base(&mut self, log: &VhostUserLog, file: File) -> VhostUserResult<()> {
+        let mem = self.atomic_mem.memory();
+
+        let logmem = Arc::new(
+            MmapLogReg::from_file(file.as_fd(), log.mmap_offset, log.mmap_size)
+                .map_err(VhostUserError::ReqHandlerError)?,
+        );
+
+        // Let's create all bitmaps first before replacing them, in case any of them fails
+        let mut bitmaps = Vec::new();
+        for region in mem.iter() {
+            let bitmap = <<T as VhostUserBackend>::Bitmap as BitmapReplace>::InnerBitmap::new(
+                region,
+                Arc::clone(&logmem),
+            )
+            .map_err(VhostUserError::ReqHandlerError)?;
+
+            bitmaps.push((region, bitmap));
+        }
+
+        for (region, bitmap) in bitmaps {
+            region.bitmap().replace(bitmap);
```

Comment:
This takes locks in turn to replace them after each other.

Do we need to guard against something concurrently using these bitmaps while racing in a following way:
```
r1: request-handler
t: This thread that replaces the bitmap
bX: bitmaps

r1 processes the a request that impacts multiple regions/bitmaps
r1 updates b0
t replaces b0
t replaces b1
r1 updates b1
```
I assume it does not matter since we probably only consider the tracking to be started after we return from `set_log_base()`?

Reply:
> I assume it does not matter since we probably only consider the tracking to be started after we return from set_log_base()?

Sadly, it's not defined in the vhost-user spec, but is what qemu does, so I assume that is the expected behavior